# Assignment: Data Wrangling
## `! git clone https://github.com/DS3001/wrangling`
## Do Q2, and one of Q1 or Q3.

# Tidy Data by Hadley Wickham - Q&A

---

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

---

### 1. Read the abstract. What is this paper about?
> This paper discusses data cleaning, particularly a niche called data tidying. Data tidying is the structuring of datasets to make manipulation, visualization, and modeling more feasible. Setting up this framework enables others to use it with much more ease because much less tools are needed to handle the data. Wickham also emphasizes the importance of needing to treat the data cleaning as a very structured process due to its impact later on.

---

### 2. Read the introduction. What is the "tidy data standard" intended to accomplish?
> The "tidy data standard" intends to structure datasets in a way that makes manipulation, visualization, and modeling easier. Tidy data follows specific formatting which makes datasets easier to process and read due to the increased uniformity. This offers clarity, usability, and efficiency to those working with the datasets due to the framework provided opposed to spending the majority of their time cleaning and preparing the dataset. 

---

### 3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
> The sentence, "Like families, tidy datasets are all alike but every messy dataset is messy in its own way," means that all tidy datasets follow the same guiding principles such as consistent column names, variable names, and null value handling. However, messy datasets can be messy in a variety of unique ways, such as inconsistent variable names, missing values, or improper structure.

> The sentence, "For a given dataset, it's usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general" means that in the context of a specific dataset, it's often easy to identify the rows (observations) and columns (variables). However, defining these terms universally is challenging because observations and variables can vary based on the dataset's purpose and structure.

---

### 4. Read Section 2.2. How does Wickham define values, variables, and observations?
> Wickham defines:
- **Values** as the actual data points contained within the dataset.
- **Variables** as the attributes or properties being measured.
- **Observations** as the individual entities or units for which measurements are recorded.

---

### 5. How is "Tidy Data" defined in section 2.3?
> "Tidy Data" is defined by the following rules:
- Each variable forms a column.
- Each observation forms a row.
- Each type of observational unit forms a table.

---

### 6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
> The five most common problems are:
- Column headers are values, not variable names.
- Multiple variables are stored in one column.
- Variables are stored in both rows and columns.
- Multiple types of observational units are stored in the same table.
- A single observational unit is spread across multiple tables.

> The data in Table 4 are messy because the column headers contain values (years) instead of variable names, which violates the tidy data principle that column headers should represent variables.

> "Melting" a dataset involves transforming it from a wide format (where variables are spread across multiple columns) to a long format (where variables are stored in a single column), making the data easier to manipulate and analyze.

---

### 7. Why, specifically, is Table 11 messy but Table 12 tidy and "molten"?
> Table 11 is messy because it stores multiple variables within a single column, breaking the tidy data rule of one variable per column. In contrast, Table 12 is tidy because each column represents a single variable and each row represents an observation. Table 12 is also "molten" because the data has been reshaped to conform to the principles of tidy data.

---

### 8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?
> The "chicken-and-egg" problem refers to the idea that tidy data is only as useful as the tools that can work with it. Since tidy data and the tools to analyze it are interdependent, progress in one area is constrained by the other. Advancements in data formats rely on corresponding advancements in data analysis tools, and vice versa.

> Wickham hopes that future research will lead to the development of new tools and frameworks for handling a wider variety of data formats. He also suggests that methodologies from cognitive science and user-centered design could improve how we design tools for data analysis, making them more intuitive and efficient.


**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)


2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.


3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.


4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [7]:
import pandas as pd

df_airbnb = pd.read_csv('./data/airbnb_hw.csv')

adjusted_values = 0
for price in df_airbnb['Price']:
    for values in price:
        if values == ',':
            adjusted_values += 1

df_airbnb['Price'] = df_airbnb['Price'].replace({',': ''}, regex=True).astype(float)

print(f"Adjusted values: {adjusted_values}")

Adjusted values: 181


My approach to cleaning the data for the airbnb data was to replace all commas with nothing in order to standardize the data. This is done using a replace statement but prior to replacing them, I loop through the 'Price' column and count all values that contain a comma, resulting in 181 'adjusted' values. However, because we do not remove any values, there are no deleted values. 

In [9]:
import numpy as np
import pandas as pd

df_sharks = pd.read_csv('./data/sharks.csv', low_memory=False)

unique_values = df_sharks['Type'].unique()
bad_values = ['Unverified', 'Invalid', 'Under investigation', 'Unconfirmed', 'Boatomg', np.nan]

df_cleaned = df_sharks[~df_sharks['Type'].isin(bad_values)].dropna(subset=['Type'])
unique_values_after = df_cleaned['Type'].unique()


Unique values after cleaning: ['Unprovoked' 'Provoked' 'Questionable' 'Watercraft' 'Boating'
 'Sea Disaster' 'Boat']


I printed the types of shark attacks, realizing that there are a lot of reports that may skew the data. These types of reports include 'Unverified', 'Invalid', 'Under investigation', 'Unconfirmed', 'Boatomg', and null values. These are all reports that aren't confirmed accurate and in such a large data set, they are not needed to draw conclusion when we have better data to base our analytics off of. This led to me filtering attacks with those 'Type' of statuses out of the report due to their questionable statuses. 

In [4]:
import numpy as np
import pandas as pd

df_pretrial = pd.read_parquet('./data/justice_data.parquet')

df_pretrial['WhetherDefendantWasReleasedPretrial'] = df_pretrial['WhetherDefendantWasReleasedPretrial'].map({'Yes': 1, 'No': 0})

df_pretrial['WhetherDefendantWasReleasedPretrial'] = df_pretrial['WhetherDefendantWasReleasedPretrial'].replace({'': np.nan, None: np.nan})


In [1]:
import pandas as pd
import numpy as np
import re

df_pretrial = pd.read_csv('../pretrial_data.csv', low_memory=False)

def clean_imposed_sentence(value):
    if pd.isna(value):
        return np.nan
    
    value = str(value).strip()
    
    if re.match(r'^\d+$', value):
        return float(value)
    
    elif re.match(r'(\d+)-(\d+)', value):
        return np.mean([float(x) for x in re.findall(r'\d+', value)])
    
    elif 'year' in value.lower():
        numbers = re.findall(r'\d+', value)
        if numbers:
            return float(numbers[0]) * 12
    
    elif 'month' in value.lower():
        numbers = re.findall(r'\d+', value)
        if numbers:
            return float(numbers[0])

    return np.nan

df_pretrial['ImposedSentenceAllChargeInContactEvent'] = df_pretrial['ImposedSentenceAllChargeInContactEvent'].apply(clean_imposed_sentence)

median_value = df_pretrial['ImposedSentenceAllChargeInContactEvent'].median()
df_pretrial['ImposedSentenceAllChargeInContactEvent'].fillna(median_value, inplace=True)

median_sentence_type = df_pretrial['SentenceTypeAllChargesAtConvictionInContactEvent'].median()
df_pretrial['SentenceTypeAllChargesAtConvictionInContactEvent'].fillna(median_sentence_type, inplace=True)

print(df_pretrial[['ImposedSentenceAllChargeInContactEvent', 'SentenceTypeAllChargesAtConvictionInContactEvent']].head())


   ImposedSentenceAllChargeInContactEvent  \
0                                     3.0   
1                                    60.0   
2                                    12.0   
3                                     3.0   
4                                     3.0   

   SentenceTypeAllChargesAtConvictionInContactEvent  
0                                                 9  
1                                                 0  
2                                                 1  
3                                                 1  
4                                                 4  


First, I converted all values under 'ImposedSentenceAllChargeInContactEvent' in order to ensure that all values are consistent and that sentences measured in years do not impact the data. If the sentence is a range of two dates, it will take the average of the bottom and top range and replace that as the value. If values are missing in the 'SentenceTypeAllChargesAtConvictionInContactEvent' variable, I replaced them with the median value to provide value to the missing data due to the size of the data. In other words, each sentence has so much meaningful data so I don't want to delete the whole row of data just because of one missing value. 

**Q3.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?